src/OpenTelemetry/Metrics/Exemplar/Exemplar.cs

```diff
@@ -28,31 +30,149 @@ namespace OpenTelemetry.Metrics;
 #endif
     struct Exemplar
 {
+    internal HashSet<string>? KeyFilter;
+    private int tagCount;
+    private KeyValuePair<string, object?>[]? tagStorage;
+    private MetricPointValueStorage valueStorage;
+    private int isCriticalSectionOccupied;
+
     /// <summary>
     /// Gets the timestamp.
     /// </summary>
-    public DateTimeOffset Timestamp { get; internal set; }
+    public DateTimeOffset Timestamp { get; private set; }
 
     /// <summary>
     /// Gets the TraceId.
     /// </summary>
-    public ActivityTraceId? TraceId { get; internal set; }
+    public ActivityTraceId? TraceId { get; private set; }
 
     /// <summary>
     /// Gets the SpanId.
     /// </summary>
-    public ActivitySpanId? SpanId { get; internal set; }
+    public ActivitySpanId? SpanId { get; private set; }
 
-    // TODO: Leverage MetricPointValueStorage
-    // and allow double/long instead of double only.
+    /// <summary>
+    /// Gets the long value.
+    /// </summary>
+    public long LongValue
+    {
+        readonly get => this.valueStorage.AsLong;
+        private set => this.valueStorage.AsLong = value;
+    }
 
     /// <summary>
     /// Gets the double value.
     /// </summary>
-    public double DoubleValue { get; internal set; }
+    public double DoubleValue
+    {
+        readonly get => this.valueStorage.AsDouble;
+        private set => this.valueStorage.AsDouble = value;
+    }
 
     /// <summary>
     /// Gets the FilteredTags (i.e any tags that were dropped during aggregation).
     /// </summary>
-    public List<KeyValuePair<string, object?>>? FilteredTags { get; internal set; }
+    public readonly ReadOnlyFilteredTagCollection FilteredTags
+        => new(this.KeyFilter, this.tagStorage ?? Array.Empty<KeyValuePair<string, object?>>(), this.tagCount);
+
+    internal void Update<T>(in ExemplarMeasurement<T> measurement)
+        where T : struct
+    {
+        if (Interlocked.Exchange(ref this.isCriticalSectionOccupied, 1) != 0)
+        {
+            // Some other thread is already writing, abort.
+            return;
+        }
+
+        this.Timestamp = DateTimeOffset.UtcNow;
+
+        if (typeof(T) == typeof(long))
+        {
+            this.LongValue = (long)(object)measurement.Value;
+        }
+        else if (typeof(T) == typeof(double))
+        {
+            this.DoubleValue = (double)(object)measurement.Value;
+        }
+        else
+        {
+            Debug.Fail("Invalid value type");
+            this.DoubleValue = Convert.ToDouble(measurement.Value);
+        }
+
+        var currentActivity = Activity.Current;
+        if (currentActivity != null)
+        {
+            this.TraceId = currentActivity.TraceId;
+            this.SpanId = currentActivity.SpanId;
+        }
+        else
+        {
+            this.TraceId = default;
+            this.SpanId = default;
+        }
+
+        this.StoreRawTags(measurement.Tags);
```

Comment:
@utpilla Co-incidentally, the active-passive buffer idea is something we discussed as well for the entire Aggregator store as well!

Reply:
>  A measuring thread could request the active buffer and still be working on completing a measurement when the collect fires. On the actual exemplar there is another flag used to detect that race as well as two threads trying to write the same exemplar

^^ This is something we can explore for agg.store as well, if it helps with cleanup/other benefits!

Reply:
Definitely! Why I'm not rushing into fixing anything on this PR for just exemplars is I want to explore if there is a way to fix it universally so that histograms can also benefit and get a lock-free update/write too.